m2cgen/ast.py

```diff
@@ -230,3 +230,33 @@ def __init__(self, expr, to_reuse=False):
     def __str__(self):
         args = ",".join([str(self.expr), "to_reuse=" + str(self.to_reuse)])
         return "SubroutineExpr(" + args + ")"
+
+
+NESTED_EXPRS_MAPPINGS = [
```

Comment:
Few things:
1. The existing implementation of the algorithm is self-contained and is not spread across dozen classes.
2. I'd really love to keep existing AST entities concise and not a pollute a data structure definition with random behavior. Moving the count logic inside will create a precedent and future temptation to throw more stuff in there.
3. With `exclude_list` parameter each entity will have to check whether it's not excluded, or otherwise check its branches. This logic will have to reside in a base class or something, which again introduces extra complexity.

Reply:
4. I'm coming from functional programming background and seeing AST as [ADTs](https://en.wikipedia.org/wiki/Algebraic_data_type), which I just realized may not be as appropriate in Python world :D

Reply:
Many thanks for the detailed explanation! It was really interesting to hear your vision.